### ISSUE TITLE
twisted.internet.endpoints.SSHCommandClientEndpoint.connect with new connections should return a cancellable Deferred

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6532]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6532|
|Type|enhancement|
|Created|2013-05-30 18:13:22Z|
|Branch|https://github.com/twisted/twisted/tree/ssh-endpoint-cancellable-6532-2|

If you are creating an SSH endpoint with a new connection each time, the `Deferred` returned for endpoint connection should be cancellable. It should call `transport.abortConnnection` on its TCP connection to abort the connection attempt if it's already connected and is in SSH protocol negotiation mode.

(This is a separate feature from cancellable reusable connections since those presumably don't want to close the shared TCP connection.)

<details><summary>Searchable metadata</summary>

```
trac-id__6532 6532
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_ssh_endpoint_cancellable_6532_2 branches/ssh-endpoint-cancellable-6532-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1369937602000000 1369937602000000
changetime__1370353526000000 1370353526000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [38503]) Branching to 'ssh-endpoint-cancellable-6532'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @itamarst<br>@itamarst set status to `assigned`|
|-|-|

I believe the necessary work is:

  1. Extend the connection helpers with a `destroyConnection` method that does `abortConnection` on the protocol transport in the case of new connections helper. Presumably it does nothing for the existing connections helper.
  2. `SSHCommandClientEndpoint._executeCommand` should call `destroyConnection` rather than `cleanupConnection` on errors.
  3. Cancelling the `factory.connectionReady` `Deferred` should again call `abortConnection` on the transport.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

> Extend the connection helpers with a destroyConnection method that does abortConnection on the protocol transport in the case of new connections helper. Presumably it does nothing for the existing connections helper

Hmm, that's too bad - considering there's already `cleanupConnection`.  Do we have to double all the interfaces now, so we have one API to ask for `loseConnection` and one API to ask for `abortConnection`? :/

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

It's necessary in so far as `cleanupConnection` is also called in other code paths which want a clean cleanup, as it were. We could add an extra argument to `cleanupConnection`, instead of adding a new method, which should be fine as long as we merge this before next release since this is new code.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [38627]) Test cancel while waiting for exec without monkey-patching an implementation detail of the endpoint

refs [#6532](https://github.com/twisted/twisted/issues/6532)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [38628]) Merge private-conn-creator-interface-6541

Author: exarkun
Reviewer: jonathanj
Fixes: [#6541](https://github.com/twisted/twisted/issues/6541)

Change `ISSHConnectionCreator` to `_ISSHConnectionCreator` so as not to make it a
part of the public interface of `twisted.conch.endpoints`, pending some changes we
probably want to make to it (at least for cancellation - see [#6532](https://github.com/twisted/twisted/issues/6532)).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [38630]) Branching to 'ssh-endpoint-cancellable-6532-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst set owner to @exarkun<br>@itamarst set status to `new`|
|-|-|

I've got a build running, haven't looked at it yet (actually two, you'll want to wait for the later one to finish).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @itamarst<br>@exarkun set status to `new`|
|-|-|

Thanks.  Just a couple points, mostly about encapsulation:

  1. `_CommandTransport` now uses a lot of attributes from its _creator_ which are not documented as part of `_ISSHConnectionCreator`.  This code will really only work with a `_NewConnectionHelper` instance.  It's true `_CommandTransport` was quite closely tied to `_NewConnectionHelper` before this change, but at least the documentation doesn't suggest otherwise.  Now it has only the creator and uses a mix of documented and undocumented attributes from it.  I'm not sure what resolution I'd like to see for this.  Maybe the interface just needs to document all of these attributes?  They're hardly well-considered though, it's just the collection of things I needed to implement the endpoint.  There are clear problems with the collection as a while - for example, the issue of having keys separate from an agent.  And stuff is missing from it, for example it cannot support keyboard-interactive authentication as is.
  1. The `connection.transport.transport.abortConnection()` necessary in the implementation of `cleanupConnection` makes me sad.  This kind of deep object usage seems to be often necessary when working with conch.  Is there something we could do to make this better?  Should the conch SSH transport class implement abortConnection to help out here?

Sorry about the general vagueness of those points.  Since this is all private stuff, it's not massively urgent to resolve it before this ticket can be closed.  It would be nice to have at least an inkling of a plan about where we're going to go from here, though.

Please merge when you're satisfied these points have been addressed _some_how, even if not necessarily by fixing all of the internals to be completely perfect.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

For the first item, how about just documenting that `_CommandTransport` only takes a `_NewConnectionHelper`, and then it's using expected public attributes?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

> For the first item, how about just documenting that _CommandTransport only takes a _NewConnectionHelper, and then it's using expected public attributes?

Sure, that works.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst commented|
|-|-|

OK, updated `_CommandTransport.__init__` docstring, and added [#6548](https://github.com/twisted/twisted/issues/6548) for point 2.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst set status to `closed`|
|-|-|

(In [38660]) Merge ssh-endpoint-cancellable-6532-2: Cancellable endpoints for new-connection SSH endpoint.

Fixes: [#6532](https://github.com/twisted/twisted/issues/6532)
Author: itamar, exarkun
Review: exarkun, itamar

If you use an SSH command endpoint that opens a new connection, you can cancel the connection attempt and SSH negotation by cancelling the connection Deferred.
